src/syscall/libc_wasip2.go

```diff
@@ -0,0 +1,1294 @@
+
+	}
+
+	streams, ok := wasiFiles[fd]
+	if !ok {
+		// TODO(dgryski): EINVAL?
+		libcErrno = EBADF
+		return -1
+	}
+	if streams.d == cm.ResourceNone {
+		libcErrno = EBADF
+		return -1
+	}
+	if streams.oflag&O_RDONLY == 0 {
+		libcErrno = EBADF
+		return -1
+	}
+
+	result := streams.d.Read(types.FileSize(count), types.FileSize(offset))
+	if err := result.Err(); err != nil {
+		libcErrno = errorCodeToErrno(*err)
+		return -1
+	}
+
+	list := result.OK().F0
+	copy(unsafe.Slice(buf, count), list.Slice())
+
+	// TODO(dgryski): EOF bool is ignored?
+	return int(list.Len())
+}
+
+// ssize_t pwrite(int fd, void *buf, size_t count, off_t offset);
+//
+//go:export pwrite
+func pwrite(fd int32, buf *byte, count uint, offset int64) int {
+	// TODO(dgryski): Need to be consistent about all these checks; EBADF/EINVAL/... ?
+	if stream, ok := wasiStreams[fd]; ok {
+		return writeStream(stream, buf, count, 0)
+	}
+
+	streams, ok := wasiFiles[fd]
+	if !ok {
+		// TODO(dgryski): EINVAL?
+		libcErrno = EBADF
+		return -1
+	}
+	if streams.d == cm.ResourceNone {
+		libcErrno = EBADF
+		return -1
+	}
+	if streams.oflag&O_WRONLY == 0 {
+		libcErrno = EBADF
+		return -1
+	}
+
+	result := streams.d.Write(cm.NewList(buf, count), types.FileSize(offset))
+	if err := result.Err(); err != nil {
+		// TODO(dgryski):
+		libcErrno = errorCodeToErrno(*err)
+		return -1
+	}
+
+	return int(*result.OK())
+}
+
+// ssize_t lseek(int fd, off_t offset, int whence);
+//
+//go:export lseek
+func lseek(fd int32, offset int64, whence int) int64 {
+	if _, ok := wasiStreams[fd]; ok {
+		// can't lseek a stream
+		libcErrno = EBADF
+		return -1
+	}
+
+	stream, ok := wasiFiles[fd]
+	if !ok {
+		libcErrno = EBADF
+		return -1
+	}
+	if stream.d == cm.ResourceNone {
+		libcErrno = EBADF
+		return -1
+	}
+
+	switch whence {
+	case 0: // SEEK_SET
+		stream.offset = offset
+	case 1: // SEEK_CUR
+		stream.offset += offset
+	case 2: // SEEK_END
+		result := stream.d.Stat()
+		if err := result.Err(); err != nil {
+			libcErrno = errorCodeToErrno(*err)
+			return -1
+		}
+		stream.offset = int64(result.OK().Size) + offset
+	}
+
+	return int64(stream.offset)
+}
+
+// int close(int fd)
+//
+//go:export close
+func close(fd int32) int32 {
+	if _, ok := wasiStreams[fd]; ok {
+		// TODO(dgryski): Do we need to do any stdin/stdout/stderr cleanup here?
+		delete(wasiStreams, fd)
+		return 0
+	}
+
+	streams, ok := wasiFiles[fd]
+	if !ok {
+		libcErrno = EBADF
+		return -1
+	}
+	if streams.d != cm.ResourceNone {
+		streams.d.ResourceDrop()
+		streams.d = 0
+	}
+	delete(wasiFiles, fd)
+
+	return 0
+}
+
+// int dup(int fd)
+//
+//go:export dup
+func dup(fd int32) int32 {
+	// is fd a stream?
+	if stream, ok := wasiStreams[fd]; ok {
+		newfd := findFreeFD()
+		wasiStreams[newfd] = stream
+		return newfd
+	}
+
+	// is fd a file?
+	if file, ok := wasiFiles[fd]; ok {
+		// scan for first free file descriptor
+		newfd := findFreeFD()
+		wasiFiles[newfd] = file
+		return newfd
+	}
```

Comment:
If you `dup()` a file FD, then `close()` one of the file descriptors, I would expect the dup-ed file descriptor to still work and point to the same `wasiFile` or `wasiStream`. Maybe you need some sort of refcount thing here both in the `wasiFile` and `wasiStream` structs so that if `refCount > 1`, they're only removed from the maps?

Reply:
I can add ref counting here.

Reply:
Ref counting added.